File: app/src/main/java/ru/kuchanov/scpquiz/controller/api/QuizApi.java
package ru.kuchanov.scpquiz.controller.api;

import java.util.List;

import io.reactivex.Single;
import retrofit2.http.Field;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.GET;
import retrofit2.http.Header;
import retrofit2.http.POST;
import ru.kuchanov.scpquiz.controller.api.response.TokenResponse;
import ru.kuchanov.scpquiz.model.api.NwQuiz;

public interface QuizApi {

    @FormUrlEncoded
    @POST("scp-quiz/oauth/token")
    Single<TokenResponse> getAccessToken(
            @Header("Authorization") String authorization,
            @Field("grant_type") String grantType
    );


    @GET("scp-quiz/quiz/all")
    Single<List<NwQuiz>> getNwQuizList(
            @Header("Authorization") String authorization
    );
}

File: app/src/main/java/ru/kuchanov/scpquiz/controller/api/QuizApi.kt
package ru.kuchanov.scpquiz.controller.api

import io.reactivex.Single
import retrofit2.http.*
import ru.kuchanov.scpquiz.controller.api.response.TokenResponse
import ru.kuchanov.scpquiz.model.api.NwQuiz
import ru.kuchanov.scpquiz.model.api.NwUser

interface QuizApi {

    /**
     * use it only for get quizList if user doesn't log in.
     */
    @FormUrlEncoded
    @POST("oauth/token")
    fun getUnloginedUserAccessToken(
            @Header("Authorization") authorization: String,
            @Field("grant_type") grantType: String
    ): Single<TokenResponse>

    /**
     * use token from [getUnloginedUserAccessToken]
     */
    @GET("quiz/all")
    fun getNwQuizList(@Header("Authorization") authorization: String): Single<List<NwQuiz>>

    @GET("user/score")
    fun getServerUserScore(): Single<Long>

    @GET("user/meClient")
    fun getNwUser(): Single<NwUser>
}

